### PR TITLE
feat(Group/AddChar): constructor for additive characters

### PR DESCRIPTION
**Pull request by loefflerd**
Currently it's a little painful to explicitly construct additive characters, because you need to use `change` etc to see through the `Multiplicative` type synonym. This adds a shortcut which directly constructs an `AddChar` from a function satisfying `f (a + b) = f a * f b` and `f 0 = 1`.

---

- [x] depends on: #11312

### CODE REVIEW

**Review by MichaelStollBayreuth — APPROVED** [review]
LGTM

**Review by YaelDillies — COMMENTED** [review]
In LeanAPAP, I have made an equivalence between `AddChar` and `AddMonoidHom`. This is functionally equivalent to what you have here, so I just wanted to mention it.

File: Mathlib/Algebra/Group/AddChar.lean [hunk 1]
```diff
@@ -61,22 +62,23 @@ namespace AddChar
...
-variable (A : Type u) [AddMonoid A] (M : Type v) [CommMonoid M]
 
-instance : CommMonoid (AddChar A M) :=
+variable (A : Type u) (M : Type v)
+
+instance [AddZeroClass A] [CommMonoid M] : CommMonoid (AddChar A M) :=
   inferInstanceAs (CommMonoid (Multiplicative A →* M))
 
-instance : Inhabited (AddChar A M) :=
+instance [AddZeroClass A] [Monoid M] : Inhabited (AddChar A M) :=
```
Comment (faenuccio) [inline Mathlib/Algebra/Group/AddChar.lean:71]: Is it really needed? I commented it and everything compiles.

File: Mathlib/Algebra/Group/AddChar.lean [hunk 2]
```diff
@@ -85,49 +87,66 @@ open Multiplicative
...
 -- Porting note(https://github.com/leanprover-community/mathlib4/issues/5229): added.
 @[ext]
-theorem ext (f g : AddChar A M) (h : ∀ x : A, f x = g x) : f = g :=
+theorem ext [Monoid M] (f g : AddChar A M) (h : ∀ x : A, f x = g x) : f = g :=
   MonoidHom.ext h
 
 /-- An additive character maps `0` to `1`. -/
 @[simp]
-theorem map_zero_one (ψ : AddChar A M) : ψ 0 = 1 := by rw [coe_to_fun_apply, ofAdd_zero, map_one]
+theorem map_zero_one [Monoid M] (ψ : AddChar A M) : ψ 0 = 1 := by
```
Comment (faenuccio) [inline Mathlib/Algebra/Group/AddChar.lean:116]:
```suggestion
theorem map_zero_eq_one [Monoid M] (ψ : AddChar A M) : ψ 0 = 1 := by
```

File: Mathlib/Algebra/Group/AddChar.lean [hunk 3]
```diff
@@ -85,49 +87,66 @@ open Multiplicative
...
-theorem map_nsmul_pow (ψ : AddChar A M) (n : ℕ) (x : A) : ψ (n • x) = ψ x ^ n := by
+theorem map_nsmul_pow [Monoid M] (ψ : AddChar A M) (n : ℕ) (x : A) : ψ (n • x) = ψ x ^ n := by
   rw [coe_to_fun_apply, coe_to_fun_apply _ x, ofAdd_nsmul, map_pow]
 #align add_char.map_nsmul_pow AddChar.map_nsmul_pow
 
 end CoeToFun
 
+section Constructor
+
+/-- Construct an `AddChar` from a function satisfying `f (a + b) = f a * f b` and `f 0 = 1`. -/
```
Comment (faenuccio) [inline Mathlib/Algebra/Group/AddChar.lean:136]: Why do you want `f` implicit here? Don't you think that a call like `mk f` would be useful?

File: Mathlib/Algebra/Group/AddChar.lean [hunk 4]
```diff
@@ -85,49 +87,66 @@ open Multiplicative
...
+    (map_add_mul' : ∀ a b : A, f (a + b) = f a * f b) (map_zero_one' : f 0 = 1) :
+    AddChar A M where
+  map_one' := map_zero_one'
+  map_mul' := map_add_mul'
+
+@[simp]
+lemma coe_mk {A M : Type*} [AddMonoid A] [Monoid M] {f : A → M}
+    (map_add_mul' : ∀ a b : A, f (a + b) = f a * f b) (map_zero_one' : f 0 = 1) :
+    (mk map_add_mul' map_zero_one' : A → M) = f := by rfl
+
```
Comment (faenuccio) [inline Mathlib/Algebra/Group/AddChar.lean:147]: Do you want to add an `ext` lemma saying that if you start from two functions satisfying the above properties and being equal on every element of `A` their `mk` coincide?

Comment (YaelDillies) [reply Mathlib/Algebra/Group/AddChar.lean:147]: I'm not sure what you mean here.

File: Mathlib/Algebra/Group/AddChar.lean (hunk 2 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Group/AddChar.lean:116]: I agree but it's out of scope for this PR

File: Mathlib/Algebra/Group/AddChar.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Group/AddChar.lean:71]: You mean the inhabited instance? We had a linter for it in Lean 3 to avoid writing void structures.

File: Mathlib/Algebra/Group/AddChar.lean (hunk 2 above)
Comment (faenuccio) [reply Mathlib/Algebra/Group/AddChar.lean:116]: Well, is it a good reason for not letting things improve little by little? Of course @loefflerd will do what he wants, but I can suggest this modification, no?

File: Mathlib/Algebra/Group/AddChar.lean (hunk 4 above)
Comment (faenuccio) [reply Mathlib/Algebra/Group/AddChar.lean:147]: It is, but I can imagine many occasions where two functions constructed possibly in different ways give rise to the same additive character and having an `ext` lemma would simplify one's life instead of repeating the above (short) proof.

File: Mathlib/Algebra/Group/AddChar.lean (hunk 2 above)
Comment (MichaelStollBayreuth) [reply Mathlib/Algebra/Group/AddChar.lean:116]: But then for consistency, you also have to change to `map_add_eq_mul`  and `map_nsmul_eq_pow` etc.
The question is whether the `eq` part serves a purpose. So far nobody seems to have been confused by its missing. (And it's `map_zero`,  `map_add` etc., not `map_zero_eq_zero` and so on.)

File: Mathlib/Algebra/Group/AddChar.lean (hunk 1 above)
Comment (MichaelStollBayreuth) [reply Mathlib/Algebra/Group/AddChar.lean:71]: It uses `inferInstanceAs`, which usually means that it is not found automatically.

File: Mathlib/Algebra/Group/AddChar.lean (hunk 4 above)
Comment (loefflerd) [reply Mathlib/Algebra/Group/AddChar.lean:147]: I'm with Yael on this one.

File: Mathlib/Algebra/Group/AddChar.lean (hunk 3 above)
Comment (loefflerd) [reply Mathlib/Algebra/Group/AddChar.lean:136]: You're right – I realised this when I went on to refactor some existing code, we need `f` explicit to be able to write `AddChar.mk (some_function) (by simp) (by simp)`

File: Mathlib/Algebra/Group/AddChar.lean (hunk 1 above)
Comment (loefflerd) [reply Mathlib/Algebra/Group/AddChar.lean:71]: This is not new code in this PR (I just weakened the typeclass assumptions slightly) so I propose to leave it as it is for now.

Comment (YaelDillies) [reply Mathlib/Algebra/Group/AddChar.lean:71]: Yes it is out of scope

File: Mathlib/Algebra/Group/AddChar.lean (hunk 2 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Group/AddChar.lean:116]: Well actually Filipo and Michael I was confused by the lack of it and made aliases for the correctly-named lemmas in LeanAPAP, but this is out of scope for the current PR.

**Review by YaelDillies — COMMENTED** [review]
Can you split off a PR for the generalisation to non-commutative monoids? It's noise in the diff and personally I would rather handle it with sections than by copying the instance assumptions to every single lemma.

TLDR that's out of scope given your PR description

**Comment by loefflerd** [comment]
> Can you split off a PR for the generalisation to non-commutative monoids? It's noise in the diff and personally I would rather handle it with sections than by copying the instance assumptions to every single lemma.
> 
> TLDR that's out of scope given your PR description

Sure

File: Mathlib/Algebra/Group/AddChar.lean (hunk 2 above)
Comment (faenuccio) [reply Mathlib/Algebra/Group/AddChar.lean]: I am still sceptical of the need of breaking a PR into two just because changing one name is "out of scope". If a significant part of a PR is about changing names or docs I would agree, but perhaps we should leave room for a bit of flexibility.

Comment (loefflerd) [reply Mathlib/Algebra/Group/AddChar.lean]: Dear Filippo, I am grateful for your suggestion, but since it is orthogonal to the original goal of the PR, I feel that as PR author I should have the option to choose not to pursue it. If you want to open a PR if your own to rename these lemmas then please go ahead

**Review by YaelDillies — COMMENTED** [review]
Upon further consideration, I think we should be even bolder and make `AddChar` its own structure.

**Comment by loefflerd** [comment]
> Upon further consideration, I think we should be even bolder and make `AddChar` its own structure.

If I understand correctly, your idea that `AddChar A M` should not be a type synonym as `Multiplicative A →* M`, but should be defined directly as a structure in its own right (with the defining properties `map_add_[eq_]mul` etc written out directly, rather than using the `Multiplicative` / `Additive` API); then we can construct an `Equiv` between this new type and `Multiplicative A →* M`, but this no longer comes for free.

I looked into this, but it turns out to be quite painful. The problem is that one can't (AFAIK) transport structure along an `Equiv`; so to prove that `AddChar A M` is a `CommMonoid` / `CommGroup` / etc under the appropriate hypotheses, one has to wade through the definition of the multiplication, proof that they're associative  has to define the multiplication, inversion, etc (& prove they satisfy the monoid axioms). This basically amounts to duplicating sizeable chunks of the existing `MonoidHom` API, which is annoying to write and will be equally annoying to maintain.

At a more philosophical level, what's the point of having the `Multiplicative` type synonym in the first place if we don't use it here, which is about the clearest place imaginable for it to be useful?

**Comment by MichaelStollBayreuth** [comment]
I remember that I had suggested a similar approach some time back, and there was a discussion on Zulip ([here](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/map.20from.20mult.20gp.20to.20add.20gp/near/271703529) and [here](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Additive.20characters/near/290005886)) where it was made clear to me that this was not the way to go. 

The reason being that one would need to duplicate all the homomorphism API for this case. Or rather multiply by a larger factor, since there should be the counterpart going from multiplicative to additive structures, and then you need 16 versions of associativity...

**Comment by YaelDillies** [comment]
We barely ever use an `AddChar` as a `MonoidHom` and when we do it's defeq abuse. Also note that not everything is coming from `MonoidHom` right now anyway. The inverse on `AddChar` is defined using the inverse on the domain, while for `MonoidHom` it's on the codomain.

**Comment by loefflerd** [comment]
>  Soon, I will make `AddChar` an additive group, and that will be one further difference with `MonoidHom`.

I would not be happy with such a change.

> Also surely you can transport structure along an `Equiv`? See eg `Logic.Equiv.Transfer`.

That's cool, I didn't know about that.

--

I have a working draft of a rewrite of `AddChar.lean` using Yaël's proposed approach (so `AddChar` is a structure in its own right but with `Equivs` to both `Multiplicative A →* M` and `A →+ Additive M`). In the process I've relaxed the typeclass assumptions considerably, and added a few useful lemmas (e.g. composition). I can post this here, but it is definitely well beyond the scope of the original PR; so maybe I should make a new one and close this one.

**Comment by MichaelStollBayreuth** [comment]
> Soon, I will make `AddChar` an additive group,

I'm against that: the convention is that the group law on a group of homomorphisms is written like that of the target group, and I don't see a reason to deviate from this here. (It's OK to have `f(x + y) = f(x)*f(y)`, but would be strange to have `(f + g)(x) = f(x)*g(x)`; this would go against all the APi we have for `Pi.blah`.)

**Comment by YaelDillies** [comment]
You will have to fight 30 years of additive combinatorics literature then 😉

Anyway, this is not a discussion for this PR

**Comment by MichaelStollBayreuth** [comment]
I guess it may be OK to have the additive group structure *in addition* to the multiplicative one (i.e., such that `add = mul`). What I'm opposed to is *removing* the multiplicative structure.

**Comment by loefflerd** [comment]
Closing this as ot is superseded by #11375 .
